lucet-runtime/src/lib.rs

```diff
@@ -133,6 +133,133 @@
+//!
+//! #[no_mangle]
+//! pub extern "C" fn run() {
+//!     unsafe {
+//!         hostcall_factorials(5);
+//!     }
+//! }
+//! ```
+//!
+//! In our hostcall, there are two changes from a standard recursive implementation of factorial.
+//!
+//! - Instead of performing the `n * fact(n - 1)` multiplication ourselves, we yield the operands
+//! and expect the product when resumed.
+//!
+//! - Whenever we have an intermediate answer, we yield it.
+//!
+//! To implement this, we introduce a new `enum` type to represent what we want the host to do next,
+//! and yield it when appropriate.
+//!
+//! ```no_run
+//! use lucet_runtime::lucet_hostcalls;
+//! use lucet_runtime::vmctx::Vmctx;
+//!
+//! pub enum FactorialsK {
+//!     Mult(u64, u64),
+//!     Result(u64),
+//! }
+//!
+//! lucet_hostcalls! {
+//!     #[no_mangle]
+//!     pub unsafe extern "C" fn hostcall_factorials(
+//!         &mut vmctx,
+//!         n: u64,
+//!     ) -> () {
+//!         fn fact(vmctx: &mut Vmctx, n: u64) -> u64 {
+//!             if n <= 1 {
+//!                 // yield an answer
+//!                 vmctx.yield_val(FactorialsK::Result(1));
+//!                 1
+//!             } else {
+//!                 let n_rec = fact(vmctx, n - 1);
+//!                 // yield a request for the host to perform multiplication
+//!                 let n = vmctx.yield_val_expecting_val(FactorialsK::Mult(n, n_rec));
+//!                 // yield an answer
+//!                 vmctx.yield_val(FactorialsK::Result(n));
+//!                 n
+//!             }
+//!         }
+//!         fact(vmctx, n);
+//!     }
+//! }
+//! ```
+//!
+//! The host side of the code, then, is an interpreter that repeatedly checks the yielded value and
+//! performs the appropriate operation. The hostcall returns normally when it is finished, so we
+//! exit the loop when the run/resume result is `Ok`.
+//!
+//! ```no_run
+//! # pub enum FactorialsK {
+//! #     Mult(u64, u64),
+//! #     Result(u64),
+//! # }
+//! use lucet_runtime::{DlModule, Error, Limits, MmapRegion, Region};
+//!
+//! let module = DlModule::load("factorials_guest.so").unwrap();
+//! let region = MmapRegion::create(1, &Limits::default()).unwrap();
+//! let mut inst = region.new_instance(module).unwrap();
+//!
+//! let mut factorials = vec![];
+//!
+//! let mut res = inst.run("run", &[]);
+//!
+//! while let Err(Error::InstanceYielded(val)) = res {
+//!     if let Some(k) = val.downcast_ref::<FactorialsK>() {
+//!         match k {
+//!             FactorialsK::Mult(n, n_rec) => {
+//!                 // guest wants us to multiply for it
+//!                 res = inst.resume_with_val(n * n_rec);
+//!             }
+//!             FactorialsK::Result(n) => {
+//!                 // guest is returning an answer
+//!                 factorials.push(*n);
+//!                 res = inst.resume();
+//!             }
+//!         }
+//!     } else {
+//!         panic!("didn't yield with expected type");
+//!     }
+//! }
+//!
```

Comment:
Trying it on for size and to make sure _I_ understand how a returned value looks vs a yielded value:
```
if let Ok(res) = res {
  factorials.push(res.into());
} else {
  panic!("instance finished with non-ok error");
};
```
(bonus: the C test already checks for the instance terminating successfully where the Rust test doesn't, so it brings them closer to testing parity)

Reply:
The functions now return the final value normally, so this is hopefully less confusing